src/util/http/HttpClient.cpp

```diff
@@ -0,0 +1,116 @@
+  if constexpr (std::is_same_v<StreamType, beast::tcp_stream>) {
+    tcp::resolver resolver{io_context_};
+    stream_ = std::make_unique<StreamType>(io_context_);
+    auto const results = resolver.resolve(host.c_str(), port.c_str());
+    stream_->connect(results);
+  } else {
+    static_assert(std::is_same_v<StreamType, ssl::stream<tcp::socket>>,
+                  "StreamType must be either boost::beast::tcp_stream or "
+                  "boost::asio::ssl::stream<boost::asio::ip::tcp::socket>");
+    ssl_context_ = std::make_unique<ssl::context>(ssl::context::sslv23_client);
+    ssl_context_->set_verify_mode(ssl::verify_none);
+    tcp::resolver resolver{io_context_};
+    stream_ = std::make_unique<StreamType>(io_context_, *ssl_context_);
+    if (!SSL_set_tlsext_host_name(stream_->native_handle(), host.c_str())) {
+      boost::system::error_code ec{static_cast<int>(::ERR_get_error()),
+                                   boost::asio::error::get_ssl_category()};
+      throw boost::system::system_error{ec};
+    }
+    auto const results = resolver.resolve(host.c_str(), port.c_str());
+    boost::asio::connect(stream_->next_layer(), results.begin(), results.end());
+    stream_->handshake(ssl::stream_base::client);
+  }
+}
+
+// ____________________________________________________________________________
+template <typename StreamType>
+std::istringstream HttpClientImpl<StreamType>::sendRequest(
+    const boost::beast::http::verb& method, const std::string& host,
+    const std::string& target, const std::string& requestBody,
+    const std::string& contentTypeHeader, const std::string& acceptHeader) {
+  // Check that we have a stream (obtained via a call to `openStream` above).
+  if (!stream_) {
+    throw std::runtime_error("Trying to send request without connection");
+  }
+
+  // Set up the request.
+  http::request<http::string_body> request;
+  request.method(method);
+  request.target(target);
+  request.set(http::field::host, host.c_str());
+  request.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+  request.set(http::field::accept, acceptHeader.c_str());
+  request.set(http::field::content_type, contentTypeHeader.c_str());
+  request.set(http::field::content_length, std::to_string(requestBody.size()));
+  request.body() = requestBody;
+
+  // Send the request, receive the response (unlimited body size), and return
+  // the body as a `std::istringstream`.
+  http::write(*stream_, request);
+  beast::flat_buffer buffer;
+  http::response_parser<http::dynamic_body> response_parser;
+  response_parser.body_limit((std::numeric_limits<std::uint64_t>::max)());
+  http::read(*stream_, buffer, response_parser);
+  std::istringstream responseBody(
+      beast::buffers_to_string(response_parser.get().body().data()));
+  return responseBody;
+}
+
+// ____________________________________________________________________________
+template <typename StreamType>
+void HttpClientImpl<StreamType>::closeStream() {
```

Comment:
I don't mind explicit `close` functions,
but shouldn't the destructor also close in case it is `forgotten`.
(Or is this case handled correctly by boost automatically without leaks and errors?)